src/coreclr/jit/valuenum.cpp

```diff
@@ -13169,10 +13169,92 @@ bool Compiler::fgValueNumberHelperCall(GenTreeCall* call)
                 // InvalidCastExc for these is set in VNForCast
                 break;
 
+            case CORINFO_HELP_READYTORUN_CHKCAST:
+            {
+                ssize_t  addrValue  = (ssize_t)call->gtEntryPoint.addr;
+                ValueNum callAddrVN = vnStore->VNForHandle(addrValue, GTF_ICON_FTN_ADDR);
+                vnpExc              = vnStore->VNPExcSetSingleton(
+                    vnStore->VNPairForFunc(TYP_REF, VNF_R2RInvalidCastExc,
+                                                        vnStore->VNPNormalPair(
+                                               call->gtArgs.GetUserArgByIndex(0)->GetNode()->gtVNPair),
+                                                        ValueNumPair(callAddrVN, callAddrVN)));
+                break;
+            }
+
+            case CORINFO_HELP_GETSHARED_GCSTATIC_BASE:
+            case CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE:
+            case CORINFO_HELP_GETSHARED_GCSTATIC_BASE_DYNAMICCLASS:
+            case CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE_DYNAMICCLASS:
+            case CORINFO_HELP_GETSHARED_GCTHREADSTATIC_BASE:
+            case CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE:
+            case CORINFO_HELP_CLASSINIT_SHARED_DYNAMICCLASS:
+            case CORINFO_HELP_GETSHARED_GCTHREADSTATIC_BASE_DYNAMICCLASS:
+            case CORINFO_HELP_GETSHARED_NONGCTHREADSTATIC_BASE_DYNAMICCLASS:
+                // These all take (Module*, class ID) as parameters.
+
+                vnpExc = vnStore->VNPExcSetSingleton(
+                    vnStore->VNPairForFunc(TYP_REF, VNF_ClassInitExc,
+                                           vnStore->VNPNormalPair(
+                                               call->gtArgs.GetUserArgByIndex(0)->GetNode()->gtVNPair),
+                                           vnStore->VNPNormalPair(
+                                               call->gtArgs.GetUserArgByIndex(1)->GetNode()->gtVNPair)));
```

Comment:
Why is it correct to give each cctor helper for the same class the same exception set, when the set of exceptions that can be thrown by a cctor (or, more precisely, the set of circumstances under which `TypeInitializationException` can be thrown) is dependent on ambient memory state as well?

(I don't necessarily think it is incorrect, since cctors are only triggered once, but am looking for the format argument of why that is, which would ideally also be included as a code comment.)

Reply:
I suppose from the perspective of VN it's hard to model something like this completely accurately. The exception set of a second call to one of these helpers is empty if it is dominated by the first call, but as you say, depends on the heap if the paths are otherwise disjoint.
So there is some aspect of "making CSE able to do something with this" here. I'll add some comments as you suggest.